File: matlab/1_code/overlay_validation.m
function overlay_validation(sweepDir, ampFig, phaseFig)
    % OVERLAY_VALIDATION Overlays both amplitude and phase results on experimental figures.
    
    if nargin < 1 || isempty(sweepDir)
        baseDir = fullfile(fileparts(mfilename('fullpath')), '..', '0_data', 'sweep_results');
        % Automatically find the most recent 'sweep_*' subfolder
        d = dir(fullfile(baseDir, 'sweep_*'));
        if isempty(d)
            % Fallback to the base directory if no subfolders exist
            sweepDir = baseDir;
        else
            [~, idx] = max([d.datenum]);
            sweepDir = fullfile(baseDir, d(idx).name);
        end
    end
    fprintf('Using sweep results from: %s\n', sweepDir);

    if nargin < 2 || isempty(ampFig)
        ampFig = fullfile(fileparts(mfilename('fullpath')), 'Figures', 'ampiezza_solo_misure_3.fig');
    end
    if nargin < 3 || isempty(phaseFig)
        phaseFig = fullfile(fileparts(mfilename('fullpath')), 'Figures', 'fase_solo_misure_3.fig');
    end

    g_cgs = 981;
    phase_rad_bath = -90 * pi / 180;

    % --- Mode Detection ---
    summaryFile = dir(fullfile(sweepDir, 'summary_*.csv'));
    isBathDriven = true; % Default
    if ~isempty(summaryFile)
        summaryData = readtable(fullfile(sweepDir, summaryFile(1).name));
        if all(summaryData.bathAmplitude_cm == 0)
            isBathDriven = false;
            fprintf('  Detected FORCING-driven sweep.\n');
        else
            fprintf('  Detected BATH-driven sweep.\n');
        end
    end

    files = dir(fullfile(sweepDir, 'CoM_gamma*_f*Hz.csv'));
    
    res = struct('gamma', [], 'freq', [], 'ampNorm', [], 'phaseDiff', []);
    count = 0;

    for k = 1:numel(files)
        tok = regexp(files(k).name, 'CoM_gamma([\d.]+)_f([\d.]+)Hz\.csv', 'tokens');
        if isempty(tok); continue; end
        g_val = str2double(tok{1}{1});
        f_val = str2double(tok{1}{2});

        opts = detectImportOptions(fullfile(sweepDir, files(k).name));
        data = readtable(fullfile(sweepDir, files(k).name), opts);
        
        t = data.time_s;
        CoM = data.CoM_cm;
        eta_boundary = data.eta_boundary_cm;

        omega = 2 * pi * f_val;
        A_ref = g_val * g_cgs / omega^2; 
        T_period = 1 / f_val;
        
        % Truncate at reflection (10% threshold)
        wave_idx = find(abs(eta_boundary) > 0.10 * A_ref, 1);
        if ~isempty(wave_idx)
            t = t(1:max(1, wave_idx-1));
            CoM = CoM(1:max(1, wave_idx-1));
        end

        n_periods_available = floor((t(end) - t(1)) / T_period);
        if n_periods_available >= 2
            count = count + 1;
            
            % --- Use only the last 3 periods for the fit ---
            n_eval = min(3, n_periods_available); 
            eval_idx = t >= (t(end) - n_eval * T_period);
            t_eval = t(eval_idx);
            CoM_eval = CoM(eval_idx);
            
            % --- Physics Mode Adjustment ---
            if isBathDriven
                z_lab = CoM_eval + A_ref * cos(omega * t_eval + phase_rad_bath);
                phase_ref_rad = phase_rad_bath;
            else
                z_lab = CoM_eval;
                phase_ref_rad = 0; % Forcing is cos(omega*t)
            end
            
            % --- Robust Least Squares Fitting ---
            M = [cos(omega * t_eval), sin(omega * t_eval), ones(size(t_eval))];
            X = M \ z_lab; 
            
            % Amplitude A = sqrt(C1^2 + C2^2)
            ampSS = sqrt(X(1)^2 + X(2)^2);
            
            % Phase of the disk: z_lab = A*cos(omega*t + phi_disk)
            phi_disk = atan2(-X(2), X(1));
            
            % Phase difference (Convention: phi_ref - phi_disk for positive lag)
            dphi = mod(rad2deg(phase_ref_rad - phi_disk), 360); % Wrap to [0, 360]
            
            res(count).gamma = g_val;
            res(count).freq = f_val;
            res(count).ampNorm = ampSS / A_ref;
            res(count).phaseDiff = dphi;
        end
    end

    gammas = [res.gamma];
    uG = unique(gammas);
    colors = lines(length(uG));
    [~, sweepName] = fileparts(sweepDir);

    %% 2. Overlay Amplitude
    if exist(ampFig, 'file')
        fprintf('Updating Amplitude Figure: %s\n', ampFig);
        figAmp = openfig(ampFig, 'new', 'visible');
        ax = gca; hold on;
        h_sim = gobjects(length(uG), 1);
        for i = 1:length(uG)
            idx = (gammas == uG(i));
            f_s = [res(idx).freq];
            a_s = [res(idx).ampNorm];
            [f_s, si] = sort(f_s); a_s = a_s(si);
            h_sim(i) = plot(ax, f_s, a_s, 'o-', 'Color', colors(i,:), 'LineWidth', 2, ...
                'MarkerSize', 8, 'MarkerFaceColor', colors(i,:), 'DisplayName', sprintf('\\Gamma = %.2f', uG(i)));
        end
        legend(ax, h_sim, 'Location', 'best', 'FontSize', 10);
        set(ax, 'YTick', 0:0.2:2.0);
        outPath = fullfile(fileparts(mfilename('fullpath')), 'Figures', sprintf('val_amp_%s', sweepName));
        saveas(figAmp, [outPath, '.fig']);
        saveas(figAmp, [outPath, '.png']);
    end

    %% 3. Overlay Phase
    if exist(phaseFig, 'file')
        fprintf('Updating Phase Figure: %s\n', phaseFig);
        figPhase = openfig(phaseFig, 'new', 'visible');
        ax = gca; hold on;
        h_sim = gobjects(length(uG), 1);
        for i = 1:length(uG)
            idx = (gammas == uG(i));
            f_s = [res(idx).freq];
            p_s = [res(idx).phaseDiff];
            [f_s, si] = sort(f_s); p_s = p_s(si);
            h_sim(i) = plot(ax, f_s, p_s, 'o-', 'Color', colors(i,:), 'LineWidth', 2, ...
                'MarkerSize', 8, 'MarkerFaceColor', colors(i,:), 'DisplayName', sprintf('\\Gamma = %.2f', uG(i)));
        end
        legend(ax, h_sim, 'Location', 'best', 'FontSize', 10);
        ylabel(ax, 'Phase Difference (deg)');
        outPath = fullfile(fileparts(mfilename('fullpath')), 'Figures', sprintf('val_phase_%s', sweepName));
        saveas(figPhase, [outPath, '.fig']);
        saveas(figPhase, [outPath, '.png']);
    end
    
    fprintf('Validation complete. Figures saved in 1_code/Figures/\n');
end

File: matlab/1_code/overlay_validation.py
import os
import sys
import glob
import re
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

def run_overlay_validation(sweep_dir=None):
    # Determine base and sweep directories
    script_dir = os.path.dirname(os.path.abspath(__file__))
    base_dir = os.path.join(script_dir, '..', '0_data', 'sweep_results')
    
    if sweep_dir is None:
        # Find the most recent 'sweep_*' folder
        sweep_folders = glob.glob(os.path.join(base_dir, 'sweep_*'))
        if not sweep_folders:
            sweep_dir = base_dir
        else:
            sweep_dir = max(sweep_folders, key=os.path.getmtime)
            
    print(f"Using sweep results from: {sweep_dir}")
    
    # Modes & Physics constants
    g_cgs = 981.0
    phase_rad_bath = -90.0 * np.pi / 180.0
    
    # Detect forcing-driven vs bath-driven
    is_bath_driven = True
    summary_files = glob.glob(os.path.join(sweep_dir, 'summary_*.csv'))
    if summary_files:
        summary_data = pd.read_csv(summary_files[0])
        if 'bathAmplitude_cm' in summary_data.columns:
            if (summary_data['bathAmplitude_cm'] == 0).all():
                is_bath_driven = False
                print("  Detected FORCING-driven sweep.")
            else:
                print("  Detected BATH-driven sweep.")
                
    # Find CoM simulation result CSVs
    sim_files = glob.glob(os.path.join(sweep_dir, 'CoM_gamma*_f*Hz.csv'))
    if not sim_files:
        print(f"No simulation CSV files found in {sweep_dir}. Aborting validation overlay.")
        return
        
    results = []
    
    for fpath in sim_files:
        fname = os.path.basename(fpath)
        match = re.match(r'CoM_gamma([\d.]+)_f([\d.]+)Hz\.csv', fname)
        if not match:
            continue
            
        g_val = float(match.group(1))
        f_val = float(match.group(2))
        
        try:
            data = pd.read_csv(fpath)
        except Exception as e:
            print(f"Error reading {fname}: {e}")
            continue
            
        if len(data) < 10:
            continue
            
        t = data['time_s'].values
        CoM = data['CoM_cm'].values
        eta_boundary = data['eta_boundary_cm'].values if 'eta_boundary_cm' in data.columns else np.zeros_like(t)
        
        omega = 2.0 * np.pi * f_val
        A_ref = g_val * g_cgs / (omega**2)
        T_period = 1.0 / f_val
        
        # Truncate at reflection (10% threshold of A_ref)
        wave_idx = np.where(np.abs(eta_boundary) > 0.10 * A_ref)[0]
        if wave_idx.size > 0:
            idx_cutoff = max(1, wave_idx[0])
            t = t[:idx_cutoff]
            CoM = CoM[:idx_cutoff]
            
        total_time = t[-1] - t[0]
        n_periods_available = int(np.floor(total_time / T_period))
        
        if n_periods_available >= 2:
            n_eval = min(3, n_periods_available)
            t_start_eval = t[-1] - n_eval * T_period
            eval_idx = t >= t_start_eval
            
            t_eval = t[eval_idx]
            CoM_eval = CoM[eval_idx]
            
            if is_bath_driven:
                z_lab = CoM_eval + A_ref * np.cos(omega * t_eval + phase_rad_bath)
                phase_ref_rad = phase_rad_bath
            else:
                z_lab = CoM_eval
                phase_ref_rad = 0.0
                
            # Least squares fitting
            M = np.column_stack([np.cos(omega * t_eval), np.sin(omega * t_eval), np.ones_like(t_eval)])
            # Solve M * X = z_lab
            try:
                X, _, _, _ = np.linalg.lstsq(M, z_lab, rcond=None)
                ampSS = np.sqrt(X[0]**2 + X[1]**2)
                phi_disk = np.arctan2(-X[1], X[0])
                dphi = np.degrees(phase_ref_rad - phi_disk) % 360.0
                
                results.append({
                    'gamma': g_val,
                    'freq': f_val,
                    'ampNorm': ampSS / A_ref,
                    'phaseDiff': dphi
                })
            except Exception as e:
                print(f"Fitting failed for {fname}: {e}")
                
    if not results:
        print("No valid steady state periods found in simulation files.")
        return
        
    df_sim = pd.DataFrame(results)
    
    # Unique gammas in simulation results
    u_gammas = sorted(df_sim['gamma'].unique())
    # Color mapping for simulation curves (matplotlib equivalent of Matlab lines)
    cmap = plt.colormaps['tab10']
    sim_colors = {g: cmap(i) for i, g in enumerate(u_gammas)}
    
    # Load experimental measurement CSVs
    figures_dir = os.path.join(script_dir, 'Figures')
    exp_data_dir = os.path.abspath(os.path.join(script_dir, '..', '0_data', 'external', 'experimental_measurements'))
    amp_csv = os.path.join(exp_data_dir, 'ampiezza_solo_misure_3.csv')
    phase_csv = os.path.join(exp_data_dir, 'fase_solo_misure_3.csv')
    
    has_exp = os.path.exists(amp_csv) and os.path.exists(phase_csv)
    if has_exp:
        df_exp_amp = pd.read_csv(amp_csv)
        df_exp_phase = pd.read_csv(phase_csv)
        exp_colors = {
            0.05: (0.6, 1.0, 0.6),
            0.20: (0.35, 0.75, 0.35),
            0.50: (0.1, 0.5, 0.1)
        }
    else:
        print("Warning: Experimental measurement CSV files not found. Simulation curves will be plotted without experimental overlay.")
        
    sweep_name = os.path.basename(sweep_dir.rstrip('/'))
    
    # --- Overlay Plot 1: Amplitude ---
    plt.figure(figsize=(8, 6))
    ax = plt.gca()
    
    # Plot experimental data in the background
    if has_exp:
        for g in sorted(df_exp_amp['gamma'].unique()):
            df_g = df_exp_amp[df_exp_amp['gamma'] == g]
            plt.errorbar(
                df_g['frequency_Hz'], 
                df_g['value'], 
                yerr=[df_g['error_lower'], df_g['error_upper']], 
                fmt='o', 
                color='k', 
                mfc=exp_colors[g], 
                mec='k', 
                ms=7, 
                capsize=3, 
                elinewidth=1.0, 
                label=f'Exp $\\Gamma = {g:.2f}$',
                zorder=2
            )
            
    # Plot simulation overlay
    for g in u_gammas:
        df_g = df_sim[df_sim['gamma'] == g].sort_values(by='freq')
        plt.plot(
            df_g['freq'], 
            df_g['ampNorm'], 
            'o-', 
            color=sim_colors[g], 
            lw=2, 
            ms=6, 
            label=f'Sim $\\Gamma = {g:.2f}$',
            zorder=3
        )
        
    plt.xlabel(r'$f$ (Hz)', fontsize=12)
    plt.ylabel(r'$A_{\mathrm{disk}} / A_{\mathrm{base}}$', fontsize=12)
    plt.title(f'Amplitude Validation Overlay ({sweep_name})', fontsize=14, fontweight='bold', pad=15)
    plt.xlim(0, 105)
    plt.ylim(0, 1.2)
    plt.xticks([0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100])
    plt.yticks([0.0, 0.2, 0.4, 0.6, 0.8, 1.0, 1.2])
    plt.grid(True, linestyle='--', alpha=0.7)
    plt.legend(loc='best', frameon=True, facecolor='white', framealpha=0.9, edgecolor='gray')
    ax.set_axisbelow(True)
    
    # Save Amplitude overlay
    plt.tight_layout()
    plt.savefig(os.path.join(figures_dir, f'val_amp_{sweep_name}.png'), dpi=300)
    plt.savefig(os.path.join(figures_dir, f'val_amp_{sweep_name}.pdf'))
    plt.close()
    
    # --- Overlay Plot 2: Phase ---
    plt.figure(figsize=(8, 6))
    ax = plt.gca()
    
    # Plot experimental data in the background
    if has_exp:
        for g in sorted(df_exp_phase['gamma'].unique()):
            df_g = df_exp_phase[df_exp_phase['gamma'] == g]
            plt.errorbar(
                df_g['frequency_Hz'], 
                df_g['value'], 
                yerr=[df_g['error_lower'], df_g['error_upper']], 
                fmt='o', 
                color='k', 
                mfc=exp_colors[g], 
                mec='k', 
                ms=7, 
                capsize=3, 
                elinewidth=1.0, 
                label=f'Exp $\\Gamma = {g:.2f}$',
                zorder=2
            )
            
    # Plot simulation overlay
    for g in u_gammas:
        df_g = df_sim[df_sim['gamma'] == g].sort_values(by='freq')
        plt.plot(
            df_g['freq'], 
            df_g['phaseDiff'], 
            'o-', 
            color=sim_colors[g], 
            lw=2, 
            ms=6, 
            label=f'Sim $\\Gamma = {g:.2f}$',
            zorder=3
        )
        
    plt.xlabel(r'$f$ (Hz)', fontsize=12)
    plt.ylabel(r'Phase Difference (deg)', fontsize=12)
    plt.title(f'Phase Validation Overlay ({sweep_name})', fontsize=14, fontweight='bold', pad=15)
    plt.xlim(0, 105)
    plt.ylim(0, 360)  # Wrapped to [0, 360]
    plt.xticks([0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100])
    plt.yticks([0, 90, 180, 270, 360])
    plt.grid(True, linestyle='--', alpha=0.7)
    plt.legend(loc='best', frameon=True, facecolor='white', framealpha=0.9, edgecolor='gray')
    ax.set_axisbelow(True)
    
    # Save Phase overlay
    plt.tight_layout()
    plt.savefig(os.path.join(figures_dir, f'val_phase_{sweep_name}.png'), dpi=300)
    plt.savefig(os.path.join(figures_dir, f'val_phase_{sweep_name}.pdf'))
    plt.close()
    
    print(f"Validation complete. Figures saved in Figures/ as val_amp_{sweep_name}.png/pdf and val_phase_{sweep_name}.png/pdf")

if __name__ == '__main__':
    sweep_dir_arg = sys.argv[1] if len(sys.argv) > 1 else None
    run_overlay_validation(sweep_dir_arg)
